File: product/src/main/kotlin/org/platypus/impl/module/product/models/ProductCategoryModel.py
class ProductCategory(models.Model):
    _name = "product.category"
    _description = "Product Category"
    _parent_name = "parent_id"
    _parent_store = True
    _parent_order = 'fieldName'
    _order = 'parent_left'

    name = fields.Char('Name', index=True, required=True, translate=True)
    parent_id = fields.Many2one('product.category', 'Parent Category', index=True, ondelete='cascade')
    child_id = fields.One2many('product.category', 'parent_id', 'Child Categories')
    type = fields.Selection([
        ('view', 'View'),
        ('normal', 'Normal')], 'Category Type', default='normal',
        help="A category of the view type is a virtual category that can be used as the parent of another category to create a hierarchical structure.")
    parent_left = fields.Integer('Left Parent', index=1)
    parent_right = fields.Integer('Right Parent', index=1)
    product_count = fields.Integer(
        '# Products', compute='_compute_product_count',
        help="The number of products under this category (Does not consider the children categories)")

    def _compute_product_count(self):
        read_group_res = self.env['product.domModule'].read_group([('categ_id', 'in', self.ids)], ['categ_id'], ['categ_id'])
        group_data = dict((data['categ_id'][0], data['categ_id_count']) for data in read_group_res)
        for categ in self:
            categ.product_count = group_data.get(categ.id, 0)

    @api.constrains('parent_id')
    def _check_category_recursion(self):
        if not self._check_recursion():
            raise ValidationError(_('Error ! You cannot create recursive categories.'))
        return True

    @api.multi
    def name_get(self):
        def get_names(cat):
            """ Return the list [cat.fieldName, cat.parent_id.fieldName, ...] """
            res = []
            while cat:
                res.append(cat.name)
                cat = cat.parent_id
            return res

        return [(cat.id, " / ".join(reversed(get_names(cat)))) for cat in self]

    @api.model
    def name_search(self, name, args=None, operator='ilike', limit=100):
        if not args:
            args = []
        if name:
            # Be sure name_search is symetric to name_get
            category_names = name.split(' / ')
            parents = list(category_names)
            child = parents.pop()
            domain = [('fieldName', operator, child)]
            if parents:
                names_ids = self.name_search(' / '.join(parents), args=args, operator='ilike', limit=limit)
                category_ids = [name_id[0] for name_id in names_ids]
                if operator in expression.NEGATIVE_TERM_OPERATORS:
                    categories = self.search([('id', 'not in', category_ids)])
                    domain = expression.OR([[('parent_id', 'in', categories.ids)], domain])
                else:
                    domain = expression.AND([[('parent_id', 'in', category_ids)], domain])
                for i in range(1, len(category_names)):
                    domain = [[('fieldName', operator, ' / '.join(category_names[-1 - i:]))], domain]
                    if operator in expression.NEGATIVE_TERM_OPERATORS:
                        domain = expression.AND(domain)
                    else:
                        domain = expression.OR(domain)
            categories = self.search(expression.AND([domain, args]), limit=limit)
        else:
            categories = self.search(args, limit=limit)
        return categories.name_get()

File: product/src/main/kotlin/org/platypus/impl/module/product/models/ProductCategoryModel.kt
package org.platypus.impl.module.product.models

import org.platypus.entity.PlatypusSelection
import org.platypus.entity.PlatypusSelectionCompanion
import org.platypus.impl.korm.model.HierarchyModel
import org.platypus.impl.korm.model.ModelNamed
import org.platypus.model.functions.PublicApiReturn
import org.platypus.impl.korm.sql.ReferenceOption
import org.platypus.impl.module.product.entities.ProductCategory

class ProductCategoryType(value: String, label: String) : PlatypusSelection<ProductCategoryModel, ProductCategory>(value, label) {
    companion object : PlatypusSelectionCompanion<ProductCategoryModel, ProductCategory, ProductCategoryType>(ProductCategoryType::class, { v, l -> ProductCategoryType(v, l) }) {
        val view by "View"
        val normal by "Normal"

        init {
            _default = normal
        }
    }
}

object ProductCategoryModel : HierarchyModel<ProductCategoryModel, ProductCategory>("product.category") {
    override val modelHelp = "Product Category"

    //    TODO _parent_name = "parent_id"
//    TODO _parent_store = True
//    TODO _parent_order = 'fieldName'
//    TODO _order = 'parent_left'
    init {
        name.extend(index = true, required = true, l10n = true)
    }

    override val parent = field.many2one("parent", label = "Parent Category", index = true,
        onDelete = ReferenceOption.CASCADE, target = ProductCategoryModel)
    override val children = field.one2many("children", label = "Child Categories", target = { ProductCategoryModel.parent })
    val type = field.selection("type", selection = ProductCategoryType,
        help = "A category of the view type is a virtual category that can be used as the parent of another category to create a hierarchical structure.")
    val productCount = field.integer("productCount", label = "# Products",
        help = "The number of products under this category (Does not consider the children categories)")

    //TODO but before read group :-)
    val computeProductCount = api.compute(productCount) {
        //        it.env.ProductTemplate.search({ProductTemplateModel.category})
        PublicApiReturn()
    }

//    TODO convert to api.constraint ? or automatique on hierarchique Model ?
//    val checkCategoryRecursion = api.private(
//        fun(self: ProductCategory) {
//            if (self.checkRecursoin()){
//                throw UserError("Error ! You cannot create recursive categories.")
//            }
//        }
//    )
}